core/src/main/scala/org/http4s/Entity.scala

```diff
@@ -16,19 +16,64 @@
 
 package org.http4s
 
-import cats.Monoid
+import fs2.{Chunk, Stream}
+import cats._
 import cats.syntax.all._
 
-final case class Entity[+F[_]](body: EntityBody[F], length: Option[Long] = None)
+sealed trait Entity[F[_]] {
```

Comment:
Thoughts on leaving `Entity` unsealed? I think we have to do this now rather than later, but two reasons:
1. We can add new entity forms later which may have a more efficient implementation. As long as they provide a `body` it should be usable by any backend.
2. Different backends may be able to process different entities in more efficient ways, with the various subtypes serving as hints. e.g. maybe some backend can serve files more efficiently than others, but we can always fall back to `body`. Furthermore, backends could define their own entity types

Reply:
So the reason I don't like an unsealed one is it means that this change improves the middleware layer not at all. As they don't know about them, so they can't leverage anything that is written in those types.

Reply:
You mean that with open types, middleware are going to have a hard time preserving those types?  With a self type and enough methods, this can be mitigated, but never entirely, and it would be uncharacteristically OOPy of us.

I guess I'd be interested in what might be missing besides a file and a HEAD.